pallets/xcmp-queue/src/lib.rs

```diff
@@ -1092,28 +1096,59 @@ impl<T: Config> XcmpMessageSource for Pallet<T> {
 	}
 }
 
+pub trait PriceForSiblingDelivery {
+	fn price_for_sibling_delivery(id: ParaId, message: &Xcm<()>) -> MultiAssets;
+}
+
+impl PriceForSiblingDelivery for () {
+	fn price_for_sibling_delivery(_: ParaId, _: &Xcm<()>) -> MultiAssets {
+		MultiAssets::new()
+	}
+}
+
+impl<T: Get<MultiAssets>> PriceForSiblingDelivery for ConstantPrice<T> {
+	fn price_for_sibling_delivery(_: ParaId, _: &Xcm<()>) -> MultiAssets {
+		T::get()
+	}
+}
+
 /// Xcm sender for sending to a sibling parachain.
 impl<T: Config> SendXcm for Pallet<T> {
-	fn send_xcm(dest: impl Into<MultiLocation>, msg: Xcm<()>) -> Result<(), SendError> {
-		let dest = dest.into();
+	type Ticket = (ParaId, VersionedXcm<()>);
 
-		match &dest {
+	fn validate(
+		dest: &mut Option<MultiLocation>,
+		msg: &mut Option<Xcm<()>>,
+	) -> SendResult<(ParaId, VersionedXcm<()>)> {
+		let d = dest.take().ok_or(SendError::MissingArgument)?;
+		let xcm = msg.take().ok_or(SendError::MissingArgument)?;
+
+		match &d {
 			// An HRMP message for a sibling parachain.
 			MultiLocation { parents: 1, interior: X1(Parachain(id)) } => {
-				let versioned_xcm = T::VersionWrapper::wrap_version(&dest, msg)
+				let id = ParaId::from(*id);
+				let price = T::PriceForSiblingDelivery::price_for_sibling_delivery(id, &xcm);
+				let versioned_xcm = T::VersionWrapper::wrap_version(&d, xcm)
 					.map_err(|()| SendError::DestinationUnsupported)?;
-				let hash = T::Hashing::hash_of(&versioned_xcm);
-				Self::send_fragment(
-					(*id).into(),
-					XcmpMessageFormat::ConcatenatedVersionedXcm,
-					versioned_xcm,
-				)
-				.map_err(|e| SendError::Transport(<&'static str>::from(e)))?;
-				Self::deposit_event(Event::XcmpMessageSent { message_hash: Some(hash) });
-				Ok(())
+				Ok(((id, versioned_xcm), price))
 			},
 			// Anything else is unhandled. This includes a message this is meant for us.
-			_ => Err(SendError::CannotReachDestination(dest, msg)),
+			_ => {
+				*dest = Some(d);
```

Comment:
We're losing `msg` content here...

```suggestion
				*dest = Some(d);
				*msg = Some(xcm);
```
Or move `let xcm = msg.take()...` to the first `match` case.